aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -52,8 +50,18 @@ pub(crate) fn validate_module_events(
         // Check all the emit calls have the correct struct with event attribute.
         validate_emit_calls(&new_event_structs, new_module)?;
 
+        // INVARIANT:
+        //   No need to charge gas for module access: this function fetches the old version of the
+        //   module which has been already charged when publish request was processed first (if
+        //   such old version exists).
+        if features.is_lazy_loading_enabled() {
+            traversal_context
+                .check_is_special_or_visited(new_module.address(), new_module.name())
+                .map_err(|err| err.finish(Location::Undefined))?;
+        }
+
         let old_module_metadata_if_exists = module_storage
-            .fetch_deserialized_module(new_module.address(), new_module.name())?
+            .unmetered_get_deserialized_module(new_module.address(), new_module.name())?
             .and_then(|module| {
                 // TODO(loader_v2): We can optimize this to fetch metadata directly.
```

Comment:
do we address this in a later PR? let's do it if we can easily, seems quite ugly with this post-processing